### PR TITLE
Update Parthenon and move to type-based variables

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

This PR was originally going to be #180. However, after making the changes described below, I decided this was a big enough changeset that it should go in first, before continuing to full sparse packs.

In this PR I move all our variables to the type-based variables. What this means is:
1. I updated to current parthenon develop, which necessitated changing some calls to `GetParentPointer` (as this API was cleaned up a little bit).
2. I added macros that declare variable types per variable and use them to declare variables in `phoebus_utils/variables.hpp`
3. Because variable names are now types, not strings, the strings must be evoked via the `variable::name()` method.

After this goes through, I will start porting each variable pack to a sparse pack, which should eliminate much of the index gymnastics going on throughout the code. 

@curiousmiah @mari2895 this is the first step in the refactor we discussed forever ago, and it should make said refactor much simpler to pursue. I also realize it'll be necessary to add face-centered constrained transport, as the topological elements are only exposed through sparse packs.

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by calling `scripts/bash/format.sh`.
- [x] Explain what you did.

### CODE REVIEW

File: src/geometry/cached_system.hpp
Comment (mari2895) [inline src/geometry/cached_system.hpp:1]: stuff here (lines 609-626) do not need name()?

**Review by mari2895 — APPROVED** [review]
@Yurlungur this is a lot of work, I appreciate it a lot! Looks good to me!

**Comment by Yurlungur** [comment]
This is an example of how the new typed variables and sparse packs can simplify code:
```C++
                                      const IndexRange &jb, const IndexRange &kb) {
  namespace p = fluid_prim;
  namespace c = fluid_cons;
  namespace impl = internal_variables;
  using parthenon::MakePackDescriptor;

  auto *pmb = rc->GetParentPointer();
  auto &resolved_pkgs = pmb->resolved_packages;

  static auto desc =
      MakePackDescriptor<p::density, c::density, p::velocity, c::momentum, p::energy,
                         c::energy, p::bfield, c::bfield, p::ye, c::ye, p::pressure,
                         p::gamma1, impl::cell_signal_speed>(resolved_pkgs.get());
  auto v = desc.GetPack(rc);

  // We need these to check whether or not these variables are present
  // in the pack. They are -1 if not present.
  const int pb_hi = v.GetUpperBoundHost(0, p::bfield());
  const int pye = v.GetUpperBoundHost(0, p::ye()); 

  auto geom = Geometry::GetCoordinateSystem(rc);
  const int nblocks = v.GetNBlocks();

  parthenon::par_for(
      DEFAULT_LOOP_PATTERN, "PrimToCons", DevExecSpace(), 0, nblocks - 1, kb.s, kb.e,
      jb.s, jb.e, ib.s, ib.e,
      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
        auto &coords = v.GetCoordinates(b);
        Real gcov4[4][4];
        geom.SpacetimeMetric(CellLocation::Cent, k, j, i, gcov4);
        Real gcon[3][3];
        geom.MetricInverse(CellLocation::Cent, k, j, i, gcon);
        Real gdet = geom.DetGamma(CellLocation::Cent, k, j, i);
        Real lapse = geom.Lapse(CellLocation::Cent, k, j, i);
        Real shift[3];
        geom.ContravariantShift(CellLocation::Cent, k, j, i, shift);

        Real S[3];
        const Real vel[] = {v(b, p::velocity(0), k, j, i), v(b, p::velocity(1), k, j, i),
                            v(b, p::velocity(2), k, j, i)};
        Real bcons[3];
        Real bp[3] = {0.0, 0.0, 0.0};
        if (pb_hi > 0) {
          for (int d = 0; d < 3; ++d) {
            bp[d] = v(b, p::bfield(d), k, j, i);
          }
        }
        Real ye_cons;
        Real ye_prim = 0.0;
        if (pye > 0) {
          ye_prim = v(b, p::ye(), k, j, i);
        }

        Real sig[3];
        prim2con::p2c(v(b, p::density(), k, j, i), vel, bp, v(b, p::energy(), k, j, i),
                      ye_prim, v(b, p::pressure(), k, j, i), v(b, p::gamma1(), k, j, i),
                      gcov4, gcon, shift, lapse, gdet, v(b, c::density(), k, j, i), S,
                      bcons, v(b, c::energy(), k, j, i), ye_cons, sig);

        for (int d = 0; d < 3; ++d) {
          v(b, c::momentum(d), k, j, i) = S[d];
        }

        if (pb_hi > 0) {
          for(int d = 0; d < 3; ++d) {
            v(b, c::bfield(d), k, j, i) = bcons[d];
          }
        }

        if (pye > 0) {
          v(b, c::ye(), k, j, i) = ye_cons;
        }

        for (int d = 0; d < 3; d++) {
          v(b, impl::cell_signal_speed(d), k, j, i) = sig[d];
        }
      });

  return TaskStatus::complete;
}

template <typename T>
TaskStatus ConservedToPrimitiveRegion(T *rc, const IndexRange &ib, const IndexRange &jb,
                                      const IndexRange &kb) {
  Mesh *pm = rc->GetMeshPointer();
  StateDescriptor *pkg = pm->packages.Get("fluid").get();
  auto c2p = pkg->Param<c2p_type<T>>("c2p_func");
  return c2p(rc, ib, jb, kb);
}

template <typename T>
TaskStatus ConservedToPrimitive(T *rc) {
  IndexRange ib = rc->GetBoundsI(IndexDomain::entire);
  IndexRange jb = rc->GetBoundsJ(IndexDomain::entire);
  IndexRange kb = rc->GetBoundsK(IndexDomain::entire);
  return ConservedToPrimitiveRegion(rc, ib, jb, kb);
}

template <typename T>
TaskStatus ConservedToPrimitiveRobust(T *rc, const IndexRange &ib, const IndexRange &jb,
                                      const IndexRange &kb) {
  // TODO(JMM): This one will not work with meshblock packs
  auto *pmb = rc->GetParentPointer();

  StateDescriptor *fix_pkg = pmb->packages.Get("fixup").get();
  auto bounds = fix_pkg->Param<fixup::Bounds>("bounds");

  StateDescriptor *pkg = pmb->packages.Get("fluid").get();
  const Real c2p_tol = pkg->Param<Real>("c2p_tol");
  const int c2p_max_iter = pkg->Param<int>("c2p_max_iter");
  const Real c2p_floor_scale_fac = pkg->Param<Real>("c2p_floor_scale_fac");
  const bool c2p_fail_on_floors = pkg->Param<bool>("c2p_fail_on_floors");
  const bool c2p_fail_on_ceilings = pkg->Param<bool>("c2p_fail_on_ceilings");
  auto invert = con2prim_robust::ConToPrimSetup(rc, bounds, c2p_tol, c2p_max_iter,
                                                c2p_floor_scale_fac, c2p_fail_on_floors,
                                                c2p_fail_on_ceilings);

  StateDescriptor *eos_pkg = pmb->packages.Get("eos").get();
  auto eos = eos_pkg->Param<Microphysics::EOS::EOS>("d.EOS");
  auto geom = Geometry::GetCoordinateSystem(rc);
  auto coords = pmb->coords;

  // TODO(JCD): move the setting of this into the solver so we can call this on MeshData
  auto fail = rc->Get(internal_variables::fail::name()).data;

  parthenon::par_for(
      DEFAULT_LOOP_PATTERN, "ConToPrim::Solve", DevExecSpace(), 0, invert.NumBlocks() - 1,
      kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
        auto status = invert(geom, eos, coords, k, j, i);
        fail(k, j, i) = (status == con2prim_robust::ConToPrimStatus::success
                             ? con2prim_robust::FailFlags::success
                             : con2prim_robust::FailFlags::fail);
      });

  return TaskStatus::complete;
}

template <typename T>
TaskStatus ConservedToPrimitiveClassic(T *rc, const IndexRange &ib, const IndexRange &jb,
                                       const IndexRange &kb) {
  using namespace con2prim;
  auto *pmesh = rc->GetMeshPointer();

  StateDescriptor *fix_pkg = pmesh->packages.Get("fixup").get();
  auto bounds = fix_pkg->Param<fixup::Bounds>("bounds");

  StateDescriptor *pkg = pmesh->packages.Get("fluid").get();
  const Real c2p_tol = pkg->Param<Real>("c2p_tol");
  const int c2p_max_iter = pkg->Param<int>("c2p_max_iter");
  auto invert = con2prim::ConToPrimSetup(rc, c2p_tol, c2p_max_iter);

  StateDescriptor *eos_pkg = pmesh->packages.Get("eos").get();
  auto eos = eos_pkg->Param<Microphysics::EOS::EOS>("d.EOS");
  auto geom = Geometry::GetCoordinateSystem(rc);
  auto fail = rc->Get(internal_variables::fail::name()).data;

  // breaking con2prim into 3 kernels seems more performant.  WHY?
  // if we can combine them, we can get rid of the mesh sized scratch array
  // TODO(JCD): revisit this.  don't think it's required anymore.  in fact the
  //            original performance thing was related to the loop being a reduce
  parthenon::par_for(
      DEFAULT_LOOP_PATTERN, "ConToPrim::Setup", DevExecSpace(), 0, invert.NumBlocks() - 1,
      kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
        invert.Setup(geom, k, j, i);
      });
  parthenon::par_for(
      DEFAULT_LOOP_PATTERN, "ConToPrim::Solve", DevExecSpace(), 0, invert.NumBlocks() - 1,
      kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
        auto status = invert(eos, k, j, i);
        fail(k, j, i) =
            (status == ConToPrimStatus::success ? FailFlags::success : FailFlags::fail);
      });
  parthenon::par_for(
      DEFAULT_LOOP_PATTERN, "ConToPrim::Finalize", DevExecSpace(), 0,
      invert.NumBlocks() - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
        invert.Finalize(eos, geom, k, j, i);
      });

  return TaskStatus::complete;
}
```

**Review by brryan — APPROVED** [review]
A huge amount of changes, a great improvement to the code though, thanks for doing this, I don't see any problems, approved.

**Pull request merged**